checks.d/sqlserver.py

```diff
@@ -307,34 +307,42 @@ def close_cursor(self, cursor):
         except Exception as e:
             self.log.warning("Could not close adodbapi cursor\n{0}".format(e))
 
-    def close_db_connections(self):
+    def close_db_connections(self, instance):
         """
         We close the db connections explicitly b/c when we don't they keep
         locks on the db. This presents as issues such as the SQL Server Agent
         being unable to stop.
         """
-        for _, connection in self.connections.iteritems():
-            try:
-                connection['conn'].close()
-            except Exception as e:
-                self.log.warning("Could not close adodbapi db connection\n{0}".format(e))
+        conn_key = self._conn_key(instance)
+        if conn_key not in self.connections:
+            return
+
+        try:
+            self.connections[conn_key]['conn'].close()
+        except Exception as e:
+            self.log.warning("Could not close adodbapi db connection\n{0}".format(e))
 
-    def open_db_connections(self):
+    def open_db_connections(self, instance):
         """
         We open the db connections explicitly, so we can ensure they are open
         before we use them, and are closable, once we are finished. Open db
         connections keep locks on the db, presenting issues such as the SQL
         Server Agent being unable to stop.
         """
-        for conn_key, connection in self.connections.iteritems():
-            conn = connection['conn']
-            timeout = connection['timeout']
-            conn_dict = {'connection_string': self._conn_string(conn_key=conn_key),
-                         'timeout': timeout}
-            try:
-                conn.connect(conn_dict)
-            except Exception as e:
-                self.log.warning("Could not connect to SQL Server\n{0}".format(e))
+
+        conn_key = self._conn_key(instance)
+        timeout = int(instance.get('command_timeout',
+                                   self.DEFAULT_COMMAND_TIMEOUT))
+        try:
+            rawconn = adodbapi.connect(self._conn_string(instance=instance),
+                                    timeout=timeout)
+            if conn_key not in self.connections:
+                self.connections[conn_key] = {'conn': rawconn, 'timeout': timeout}
+            else:
+                # Existing Connection object will be closed (if open) when GC'd
+                self.connections[conn_key]['conn'] = rawconn
```

Comment:
I don't really know how adodbapi's connections behave exactly, but given that we used to make sure to always close the connections explicitely (and not wait for GC to take care of it), we should probably explicitely close the old connection here before replacing it with the new one.